src/main/kotlin/com/exactpro/th2/common/schema/factory/LazyProvider.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.exactpro.th2.common.schema.factory
+
+import com.exactpro.th2.common.schema.exception.CommonFactoryException
+import com.exactpro.th2.common.schema.factory.LazyProvider.ThrowableConsumer
+import java.util.concurrent.Callable
+import java.util.concurrent.atomic.AtomicReference
+
+internal class LazyProvider<T : Any?> private constructor(
+    private val name: String,
+    private val supplier: Callable<out T>,
+    private val onClose: ThrowableConsumer<in T>
+) : AutoCloseable {
+    private val reference = AtomicReference<State<T>?>()
+
+    init {
+        require(name.isNotBlank()) { "blank name provided" }
+    }
+
+    fun get(): T {
+        return getState().value
+    }
+
+    /**
+     * The difference from [get] method is that this one does not throw an exception
+     * if provider already closed.
+     * However, it will still throw an exception if value initialization is failed
+     */
+    fun getOrNull(): T? {
+        return getState().takeIf { it is State.Hold }?.value
+    }
+
+    private fun getState(): State<T> {
+        return if (reference.compareAndSet(null, State.Init)) {
+            val holder = State.Hold(initialize())
+            if (!reference.compareAndSet(State.Init, holder)) {
+                onClose.consume(holder.value)
+                error("provider '$name' already closed")
+            }
+            holder
+        } else {
+            var currentState: State<T>?
+            do {
+                currentState = reference.get()
+            } while (currentState == State.Init || currentState == null)
+            currentState
+        }
+    }
+
+    @Throws(Exception::class)
+    override fun close() {
+        if (reference.compareAndSet(null, State.Closed)) {
+            // factory was not yet initialized. Just return
+            return
+        }
+
```

Comment:
I think we should add State.Init handling

Reply:
This is handled in `get` method already

Reply:
Sorry, but I don't see `get` method call in this method

Reply:
Because we don't need to call it here. If somebody calls `get` and `close` method concurrently and `close` method is invoked when the current state is `Init` the `get` method will handle resource closing because the resource is not yet created to close it in `close` method